src/Model.php

```diff
@@ -2131,6 +2131,142 @@ public function hasField($name)
 
     // }}}
 
+    // {{{ References
+
+    /**
+     * Adds injected model - one record sub_model.
+     *
+     * It will be saved as serialized field value inside system field of this model.
+     *
+     * @param string      $link
+     * @param array|Model $defaults
+     *
+     * @throws Exception
+     * @throws \atk4\core\Exception
+     *
+     * @return $this
+     */
+    public function containsOne($link, $defaults = [])
+    {
+        // prepare defaults, 'model' key will contain sub-model class/object
+        if (!is_array($defaults)) {
+            $sub_model = $defaults ?: 'Model_'.$link;
+        } elseif (isset($defaults[0])) {
+            $sub_model = $defaults[0];
+            unset($defaults[0]);
+        } else {
+            throw new Exception(['Model is not set for containsOne method', 'model' => get_class($this)]);
+        }
+
+        // first let's add hidden system field which will store this sub-model data
+        $this->addField($link, ['type' => 'array', 'system' => true, 'serialize' => 'json']);
+
+        // then define custom reference and hooks
+        $this->addRef($link, function ($self) use ($link, $sub_model, $defaults) {
+
+            // model should be loaded
+            if (!$self->loaded()) {
+                throw new \atk4\core\Exception(['Model should be loaded!', 'model'=>get_class($self)]);
+            }
+
+            // create sub model object
+            $sub_model = $self->factory($sub_model, $defaults);
+            $sub_model->id_field = null;
+            $sub_model->table = $link;
+
+            // set some hooks for sub_model
+            $sub_model->addHook(['beforeSave'], function ($m) use ($self, $link) {
+                $self->save([$link => $m->get()]);
+                $m->breakHook(false);
+            });
+
+            $sub_model->addHook(['beforeDelete'], function ($m) use ($self, $link) {
+                $self->save([$link => null]);
+                $m->breakHook(false);
+            });
+
+            // extract data into array
+            $data = $self[$link] ?: [];
+
+            /**/
+            // associate with array persistence
+            $data = [$sub_model->table => ($data ? [1 => $data] : [])];
+            $persistence = new Persistence_Array($data);
+            $persistence->add($sub_model);
+
+            // try to load any (actually only one possible) record
+            $sub_model->tryLoadAny();
```

Comment:
containsOne model don't have id_field so... i think it's better to tryLoadAny in this case.

Reply:
Also we don't have pure JSON Persistence in working condition. Probably instead of Persistence_Array we could use Persistence_JSON at some point here.